debian-bo-master/bo/source/graphics/tgif-3.0p5/tgif.man
.\"@(#)$Header: /n/opus/u/guest/william/src/tgif/v3/RCS/tgif.man,v 3.0 1996/05/06 16:12:23 william Exp $
.TH TGIF n "Version 2.16 Patchlevel 11 and Above" "Tgif"
.\"
.SH NAME
.\"
\fItgif\fR \- Xlib based interactive 2-D drawing facility under X11.
Supports hierarchical construction of drawings and easy navigation between
sets of drawings.  It can also be used to launch applications.  It's also
a hyper-graphics (or hyper-structured-graphics) browser on the World-Wide-Web.
.\"
.SH SYNOPSIS
.\"
.B tgif
[\fB-display\fR displayname]
[\fB-fg\fR <color>]
[\fB-bg\fR <color>]
[\fB-bd\fR <color>]
[\fB-rv\fR]
[\fB-nv\fR]
[\fB-bw\fR]
[\fB-cwo\fR]
[\fB-hyper\fR]
[\fB-geometry\fR <geom>]
[\fB=<geom>\fR]
[\fI<file>\fR[.\fIobj\fR]]
.PP
or
.PP
.B tgif
\fB-print\fR
[\fB-eps\fR]
[\fB-p\fR]
[\fB-ps\fR]
[\fB-f\fR]
[\fB-text\fR]
[\fB-gray\fR]
[\fB-adobe\fR | -adobe=<number>/<number>]
[\fB-page <number>\fR]
[\fB-print_cmd "<command>"\fR]
[\fB-one_file_per_page\fR]
[\fB-pepsc\fR]
[\fB-bop_hook "<string>"\fR]
[\fB-eop_hook "<string>"\fR]
[\fB-o\fRdir]
[\fIfile1\fR \fIfile2\fR ...]
.\"
.SH DESCRIPTION
.\"
\fITgif\fR is an interactive drawing tool that allows the user to draw and
manipulate objects in the X Window System.  Tgif runs interactively in
the first form.  In the second form, tgif just prints files without
opening windows or fonts, and it is compatible with
the obsoleted \fIprtgif\fR.
.PP
In the first form, the command line arguments
\fI-fg\fR, \fI-bg\fR, and \fI-bd\fR specify
the foreground, background, and border colors, respectively.
The command line argument \fIfile\fR
specifies a file or an URL (only Uniform Resource Locator that specifies
a file on a HTTP or FTP server are supported) of
objects to be initially edited by tgif.
(For a more detailed description of URL, the reader is referred to the
World-Wide-Web page, titled "A Beginner's Guild to HTML", located at
<URL:http://www.ncsa.uiuc.edu/General/Internet/WWW/HTMLPrimer.html>.)
If \fI-rv\fR (or \fI-nv\fR) is specified, tgif will come up
in reverse-video (or normal-video) mode.
If \fI-bw\fR is specified, tgif will come up in the black and white mode.
If \fI-cwo\fR is specified, only the canvas window (see TGIF SUBWINDOWS
section below) will be displayed.
If \fI-hyper\fR is specified tgif will be started in the \fIhyperspace\fR mode
(see HYPERSPACE section below).
This has the same effect as setting the Tgif*CanvasWindowOnly X default to true.
The default is false.
Tgif is purely based on \fIXlib\fR.
It is tested under X11R6, and
it requires a 3 button mouse.
.PP
In the second form, tgif prints \fIfile1.obj\fR, \fIfile2.obj\fR,
etc.  (generated by \fItgif\fR)
into PostScript(TM) page description files
and pipes them to \fIlpr\fR if none of the
\fB-eps\fR, \fB-p\fR, \fB-ps\fR, \fB-f\fR, or \fB-text\fR options
are not specified;
in this case, any other command line options are sent to \fIlpr\fR.
A symbol file (see descriptions below) can also be printed by
specifying the \fI.sym\fR extension explicitly.
Only non-color printing is supported.
.PP
In the second form,
if the \fB-eps\fR or the \fB-p\fR option is specified,
tgif generates an Encapsulated PostScript file in \fIfile.eps\fR;
this file is supposed to be included in a LaTeX file through the
\\psfig, \\epsf, or \\psfile construct (see the LATEX FIGURE FORMATS section
below).
If the \fB-ps\fR or the \fB-f\fR option is specified,
it generates a PostScript file in \fIfile.ps\fR;
this file can be printed to a PostScript printer with lpr.
If the \fB-text\fR option is specified,
it generates a text file in \fIfile.txt\fR;
the text file contains all visible text and can be fed to a spell checker.
(Note that if any of the \fB-eps\fR, \fB-p\fR, \fB-ps\fR,
\fB-f\fR, or \fB-ftext\fR options are specified,
any other command line options are ignored.)
Using the \fI-gray\fR option has the same effect as setting the
Tgif*UseGrayScale X default to true (see the X DEFAULTS section below).
Using the \fI-adobe\fR or the \fI-adobe=<number>/<number>\fR
option has the same effect as specifying the
Tgif*UsePsAdobeString X default.
Using the \fI-page\fR option causes a specified page to be printed.
Using the \fI-print_cmd\fR option has the same effect as specifying the
Tgif*PrintCommand X default.
Using the \fI-one_file_per_page\fR option causes each page to be
printed into a separate file.
Using the \fI-pepsc\fR (PreserveEPSComment) option has the same effect as
setting the Tgif*StripEPSComment X default to false.
Using the \fI-bop_hook\fR and -fI-eop_hook\fR options have the same effect
as specifying the Tgif*PSBopHook and Tgif*PSEpsHook X defaults.
If the \fI-o\fR option is not specified, the output file (eps, ps, or txt)
goes into the same directory as the input file.  If \fI-o\fRdir
is specified, the output file goes into the directory specified
by \fIdir\fR.
.PP
Please note that only running tgif interactively can generate X11 bitmap,
X11 pixmap, GIF, or color PostScript files.
.\"
.SH BASIC FUNCTIONALITIES
.\"
Primitive objects supported by tgif
are rectangles, ovals, rounded-corner rectangles, arcs,
polylines, polygons, open-splines, closed-splines, text, X11 bitmaps,
some specific forms of X11 pixmaps, and Encapsulated PostScript(TM).
Objects can be grouped together to form a \fIgrouped\fR object.
A primitive or a grouped object can be made into an \fIicon\fR object or a
\fIsymbol\fR object through user commands.
.PP
Tgif objects are stored in two types of files.  A file with a
.I \.obj
extension (referred to as an \fIobject\fR file)
is a file of objects, and a file with a
.I \.sym
extension (referred to as a \fIsymbol\fR file)
specifies a ``building-block'' object.
A \fIteleport\fR mechanism is provided to \fItravel\fR among the .obj files.
A building-block object consists of the
\fIrepresentation\fR part and the \fIdefinition\fR
part (which can be empty) of the object.
Tgif supports the ``bottom-up'' construction of hierarchical
drawings by providing the capability to ``instantiate'' a
building-block object in a drawing.  Tgif also supports the ``top-down''
specification of drawings by allowing the user to make any
object a \fIrepresentation\fR
of an un-specified subsystem.
Both types of files are stored in the
form of Prolog facts.  Prolog code can be written to interpret
the drawings!  (It is left to the user to produce the code.
See the PROLOG/C TESTDRIVE section for more details.)
Prolog engines will be referred to as
\fIdrivers\fR in the sections to follow.  (Other types of drivers
are also allowed, e.g., written in C.)
.PP
Text based \fIattributes\fR can be attached to any non-text object.
Attributes specified in the representation part of a
building-block object are non-detachable when such an object is instantiated.
See ATTRIBUTES section for details.
.PP
Tgif can generate output in four different formats.
By default, the output is in the PostScript(TM) format (color PostScript
is supported), and it is generated into a file named
/tmp/Tgifa* (produced by mktemp() calls) where * is a number;
this file is piped to lpr.  This takes place when the
laser-printer icon is displayed in the Choice Window (see below
for the naming of tgif windows).
This output can be redirected to a file with a
.I \.ps
extension.  This takes place when the \fIPS\fR icon is displayed.
When the \fILaTeX\fR (or \fIEPSI\fR) icon is displayed, the output
is generated into a file with a
.I \.eps
extension.  This file is in the Encapsulated PostScript (or Encapsulated
PostScript Interchange) format;
it can be included in a LaTeX document with the
\fI\\psfig\fR or the \fI\\epsf\fR construct; this will be discussed later.
The only difference between the EPS and EPSI formats is that an EPSI
file contains a preview bitmap.  However, it takes time to generate the
preview bitmap.  If the EPS/EPSI file is to be incorporated into some
tool that does not know how to use the preview bitmap, time can be saved
by not using the EPSI format.
When the \fIT\fR icon is displayed, the output generated into a file with a
.I \.txt
extension.  This is a text file containing all visible text;
it can be fed to a spell checker.
When the \fIx11bm\fR (X11 bitmap) icon is displayed in the
Choice Window and color output is \fInot\fR selected,
tgif generates the output with the
.I \.xbm
extension; the output is in the X11 bitmap format.
However, if the x11bm icon is displayed in the
Choice Window and color output \fIis\fR selected (through the ^#k
keyboard command -- ^ denotes the <Control> and # denotes the <Meta> key),
then tgif generates the output with the
.I \.xpm
extension, and the output is in the X11 pixmap format
(the version of this XPM format depends on the settings of the
XPmOutputVersion X default).
X11 bitmap files, certain forms of X11 pixmap files
(such as the one generated by tgif; see the section on X11 PIXMAP
for details), GIF files, and Encapsulated PostScript (EPS) files
can be \fIimported\fR into tgif and be represented
as tgif primitive objects.
.PP
Tgif drawings are supposed to be printed on letter size paper (8.5in by 11in).
Both landscape and portrait page styles are supported by tgif.
Reduction (or magnification) can be controlled by the #% keyboard
command to set the reduction/magnification.  If the compiler
flag -DA4PAPER is defined (in Imakefile or Makefile.noimake),
then the output is supposed to be printed on A4 papers (which
has approximate dimensions of 8.25in by 11.7in).
.\"
.SH GRAPHICAL OBJECTS
.\"
An object in an \fIobject\fR (\fI.obj\fR) file can be a primitive object,
a grouped object, or an \fIicon\fR object.
A \fIsymbol\fR (\fI.sym\fR) file can have any number of objects allowed in an
object file and exactly one \fIsymbol\fR object.
(Recall that a symbol file specifies a building-block
object.)  The symbol object in a symbol file is the representation
part of the building-block object, and the rest of the
symbol file is the definition
part of the building-block object.  The symbol object is highlighted
with a dashed outline to distinguish it from the rest of
the objects.  When a building-block object is instantiated,
the symbol part of the file is copied into the graphics editor, and it
becomes the icon for the building-block object.
.PP
All objects in tgif can be moved, duplicated, deleted, rotated, 
and flipped.  (However, flipping text objects horizontally will
cause the text justification to change, and flipping text objects
vertically will usually cause the text object to move.)
All objects, except text and rigid icon objects, can be stretched
(scaled).  (See the TGIF SUBWINDOWS section for the definition of rigid
icon objects.)
.PP
Tgif supports 32 fill patterns, 32 pen patterns, 7 default line widths, 4 line
styles (plain, head arrow, tail arrow, double arrows) for
polylines and open-splines, 9 dash patterns,
3 types of text justifications, 4 text
styles (roman, italic, bold, bold-italic), 11 text sizes
(8, 10, 12, 14, 18, and 24 for the 75dpi fonts and
11, 14, 17, 20, 25, and 34 for the 100dpi fonts), 5 fonts (Times, Courier,
Helvetica, New-Century-Schoolbook, Symbol), and 11 default colors (magenta,
red, green, blue, yellow, pink, cyan, cadet-blue, white, black,
dark-slate-gray).
Only right-angle rotations are supported.
.PP
Most commands in tgif can either be activated by a
popup menu or by typing an appropriate
non-alphanumeric key.  All operations that change any object can be undone
and the redone.  Commands such as zoom, scroll, change fonts while no
text objects are selected, etc. are not undoable.  The undo/redo
history buffer size can be set using the Tgif*HistoryDepth X default.
.\"
.SH TGIF SUBWINDOWS
.\"
The tgif windows are described below.
.TP
.I Top Window
Displays the current domain and the
name of the file tgif is looking at.  Mouse clicks
and key presses have no effect.
.TP
.I Menubar Window
This window is right under the Top Window.
Pull-down menus can be activated from it with any mouse buttons.
Key presses have no effect.  If HideMenubar() is selected from the Layout Menu,
this window becomes invisible.  If ShowMenubar() is selected from the
Layout Menu (which can be activated from the Canvas Window below), this
window becomes visible.
.PP
.RS
The View, Text, and Graphics pull-down menus are cascading menus and can
not be \fIpinned\fR (see the \fIPopup Menus\fR subsection below for a
description).
.RE
.TP
.I Message Window
This is right under the Menubar Window on the left.
It displays tgif messages.
Clicking the left mouse button
in this window scrolls
the messages towards the bottom, clicking the right mouse
button scrolls towards
the top, and clicking or dragging the middle mouse button scrolls
to the location in the
message history depending on where the mouse is clicked.
If the <Shift> (or <Control>) key is held down when clicking the
left/right mouse button, it scrolls right/left.
.TP
.I Panel (Choice) Window
This is the window to the right of the Message Window, and it
contains a collection of icons (not to be confused with the tgif icon objects)
reflecting the current state of tgif.
In top/bottom, left/right order, it
displays the current drawing mode, the
page style (portrait/landscape), edit (see below),
print mode, zoom factor, constrained/unconstrained move (and stretch) mode,
radius for rounded-corner rectangles,
text rotation, page number or row/column, page layout mode
(stacked or tiled), horizontal alignment (L C R S -),
vertical alignment (T M B S -),
font, text size, vertical spacing between lines of text within
the same text object, text justification, dash pattern,
line style, polyline or spline, line width, fill pattern, pen pattern,
color, and special (see below).
Key presses have no effect in this window.
.PP
.RS
In addition to displaying the current state of tgif, the icons in the
Choice Window can also be used to change the current state.
Each icon is associated with a particular state variable of tgif.
Clicking the left mouse button on top of an icon cycles the
state variable associated with the icon forward; clicking the right mouse
button cycles the state variable backwards.  Dragging the middle mouse button
on top of an icon usually generates a popup menu which corresponds to
an entry in the Main Menu for the Canvas Window below.
(The ``edit'' and ``special'' icons mentioned above are dummy icons
that allow the ``edit'' and ``special'' menus to be
accessed in the Choice Window.  They do not respond to left and right
mouse clicks.)
The response to the dragging of the middle mouse button is different for
the zoom, radius, and vertical spacing icons.  Dragging the mouse
left or up increases the zoom or decreases the radius or vertical spacing;
dragging the mouse right or down has the opposite effect.
.RE
.PP
.RS
If there are objects selected in the
canvas window, then the action of the mouse will cause the selected
objects to change to the newly selected mode; note that in this
case, the current choice won't change if the middle mouse button
is used.
.RE
.PP
.RS
The settings of the horizontal and vertical
alignments determine how objects (or vertices) align with
each other when the ^l keyboard command is issued,
how each individual object (or vertex) aligns with the grids
when the ^t keyboard command is issued,
how objects or vertices distribute spatially with respect to each other when
the #l keyboard command is issued,
and how each icon replaces the old icon when the ^#u
keyboard command is issued.  The horizontal
alignments are left (L), center (C), right (R), space (S), and ignore (-).
The vertical alignments are top (T), middle (M), bottom (B), space (S), and
ignore (-).  In aligning operations, the space (S) and the ignore (-)
settings have the same effect.  The space settings are used to distribute
objects such that the gaps between any two neighboring objects are equal.
In vertex mode, any non-ignore setting will cause the selected vertices
to be spaced out evenly.
The best way to understand them is to try them out.
.RE
.PP
.RS
The text vertical spacing determines the vertical distance to advance
when a carriage return is pressed during text editing.  If the user
tries to set the
value too negative, such that the next line is exactly at
the same position as the current line, such a setting will not be
allowed (this distance depends on the current font and font size).
.RE
.TP
.I Canvas Window
This is the drawing area.  The effects of the actions of the mouse is
determined by the current drawing mode.  Normally, dragging the right mouse
button will generate the Mode Menu.  The drawing modes are (in order,
as they appear in the Mode Menu) select, text, rectangle,
oval, polyline (open-spline), polygon (closed-spline),
arc, rounded-corner rectangle, freehand polyline (open-spline),
and select vertices.  When drawing a rectangle, an oval, or a rounded-corner
rectangle, if the <Shift>
key is held down, a square, a circle, or a rounded-corner square is drawn.
Dragging the middle mouse button will generate the Main Menu.
.PP
.RS
In the select mode, left mouse button selects, moves, stretches,
and reshapes objects (double-click will ``de-select'' all selected objects
in vertex mode).  When an object is selected, it is highlighted by
little squares at the corners/vertices (using the Tgif*HandleSize X default,
the sizes of the handles can be customized).  Dragging one of the little squares
stretches/reshapes the selected object.  If one wants to move the selected
object, one should not drag the little squares.  Instead, one should
drag other parts of the object.  For example, if the object is a
hollow rectangle (the fill is NONE and the pen is not NONE), in order to select
the rectangle, one would need to click on the outline of the rectangle
with the left mouse button.  If one would like to move the rectangle, one
would need to drag the outline of the rectangle with the left mouse button.
If the object is a filled rectangle (fill is not NONE), the one can
click inside the rectangle to select it and
drag anywhere inside the rectangle to move it.
.RE
.PP
.RS
Holding down the <Shift> key and clicking the left mouse
on an object which is not currently selected will add the
object to the list of already selected objects.  The same
action applied to an object which is already selected will cause
it to be de-selected.
When stretching objects (not reshaping poly-type objects),
holding down the <Shift> key \fIafter\fR stretching is initiated results in
proportional stretching being activated (basically, a scale operation
is being performed).
Text and rigid icon objects can not be stretched or scaled.
(Rigid icon objects are icons that do not have an inherited
attribute whose name is empty and whose value is the string "not_rigid".
Rigid icon objects inside a non-rigid icon object are considered non-rigid.)
Clicking the middle mouse button
while the <Shift> key is held down will activate
the \fIteleport\fR (or \fItravel\fR), the \fIlaunch\fR,
or the \fIexecute internal command\fR mechanism.  See the sections on
TELEPORT, LAUNCH APPLICATIONS, and INTERNAL COMMANDS for details.
Teleporting has precedence over launching, which has precedence over
executing an internal command.
The arrow keys can also be used to move selected objects.  However,
if no objects
are selected, using the arrow keys will scroll the drawing area by a small
amount, and using the arrow keys when <Control> key is held down will
scroll a screen full.
.RE
.PP
.RS
In the select vertices mode, left mouse button selects and moves
vertices.  Only the top-level polyline/open-spline and polygon/closed-spline
objects which are selected when the vertex mode is activated
are eligible for vertex operations.  In this mode, all
eligible objects have their vertices highlighted with squares.
When a vertex is selected (using similar mechanism as selecting
objects described above), it is doubly highlighted with a '+' sign.
Operations available to these doubly highlighted vertices are move, delete,
align (with each other), distribute (space them equally), and align to grid.
The arrow keys can also be used to move selected vertices.
.RE
.PP
.RS
Objects can be locked (through the #< command).  Locked object are shown
with gray handles, and they can not be moved, stretched, flipped,
or rotated.  When objects are grouped, the resulting grouped object
will also be locked if any one of it's constituents is locked.
Locked objects can have their properties, such as color, font, pen, etc.,
changed; furthermore, they can be deleted.
.RE
.PP
.RS
If the current move/stretch mode is of the constrained type (activated
and deactivated by the #@ command), top-level polylines will have the
following behavior.  In a move operation, if both endpoints of a
polyline lie inside the objects being moved, then the whole polyline
is moved; otherwise, if only one endpoint falls inside the objects
being moved, then that endpoint is moved.  The vertex that is the
neighbor of the moved endpoint may also be moved either horizontally

debian-bo-master/bo/source/graphics/tgif-3.0p5/tgif.c
/*
 * Author:      William Chia-Wei Cheng (william@cs.ucla.edu)
 *
 * Copyright (C) 1990-1996, William Chia-Wei Cheng.
 *
 * Permission limited to the use, copy, display, distribute without
 * charging for a fee, and produce derivative works of "tgif" and
 * its documentation for not-for-profit purpose is hereby granted by
 * the Author, provided that the above copyright notice appears in
 * all copies made of "tgif" and that both the copyright notice
 * and this permission notice appear in supporting documentation,
 * and that the name of the Author not be used in advertising or
 * publicity pertaining to distribution of the software without
 * specific, written prior permission.  The Author makes no
 * representations about the suitability of this software for any
 * purpose.  It is provided "as is" without express or implied
 * warranty.  All other rights (including, but not limited to, the
 * right to sell "tgif", the right to sell derivative works of
 * "tgif", and the right to distribute "tgif" for a fee) are
 * reserved by the Author.
 *
 * THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS,
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, INDIRECT
 * OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
 * LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
 * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
#ifndef lint
static char RCSid[] =
      "@(#)$Header: /n/opus/u/guest/william/src/tgif/v3/RCS/tgif.c,v 3.0 1996/05/06 16:12:21 william Exp $";
#endif

#include <stdio.h>
#include <string.h>
#include <X11/Xlib.h>
#include "const.h"
#include "types.h"

#include "color.e"
#include "file.e"
#include "mainloop.e"
#include "msg.e"
#include "obj.e"
#include "page.e"
#include "setup.e"
#include "util.e"

/*
 * extern int	malloc_debug ARGS_DECL((int));
 */

int	lastFile=TRUE;

short pDrawFontAsc[] = {
    8, 10, 12, 14, 17, 22, 
    8, 10, 12, 14, 17, 22, 
    8, 10, 12, 14, 17, 23, 
    8, 10, 12, 14, 17, 22, 
    8,  9, 11, 13, 15, 19, 
    7,  9, 11, 12, 15, 20, 
    7,  9, 11, 13, 15, 19, 
    7,  9, 11, 12, 15, 20, 
    9, 11, 12, 14, 18, 24, 
    9, 11, 12, 14, 17, 24, 
    9, 11, 12, 14, 17, 24, 
    9, 11, 12, 14, 17, 24, 
    8, 11, 12, 14, 18, 23, 
    8, 11, 12, 15, 18, 24, 
    8, 11, 12, 14, 16, 23, 
    8, 11, 12, 14, 16, 24, 
    8, 10, 12, 14, 18, 24, 
    9, 14, 15, 18, 23, 30, 
   10, 14, 16, 17, 23, 30, 
    9, 13, 15, 17, 22, 30, 
    9, 13, 15, 18, 22, 30, 
    9, 12, 14, 15, 19, 26, 
    9, 12, 14, 15, 21, 26, 
    9, 12, 14, 15, 19, 26, 
    9, 12, 14, 15, 20, 26, 
   11, 14, 16, 18, 24, 31, 
   11, 14, 16, 18, 24, 31, 
   11, 14, 16, 18, 24, 31, 
   11, 14, 16, 18, 24, 31, 
   11, 14, 16, 19, 24, 32, 
   11, 15, 16, 19, 24, 33, 
   11, 14, 16, 18, 23, 32, 
   11, 15, 16, 19, 24, 32, 
   11, 12, 13, 14, 19, 27 
};
short pDrawFontDes[] = {
    2,  3,  3,  4,  4,  6, 
    2,  3,  3,  4,  4,  6, 
    2,  3,  3,  4,  5,  6, 
    2,  3,  3,  3,  4,  6, 
    2,  2,  3,  3,  4,  5, 
    2,  2,  3,  4,  5,  5, 
    2,  2,  3,  3,  4,  5, 
    2,  2,  3,  4,  5,  5, 
    2,  2,  3,  3,  4,  5, 
    2,  2,  3,  3,  5,  5, 
    2,  2,  3,  3,  5,  5, 
    2,  2,  3,  3,  5,  5, 
    2,  2,  3,  3,  4,  5, 
    2,  2,  3,  3,  4,  5, 
    2,  2,  3,  3,  6,  5, 
    2,  2,  3,  3,  6,  5, 
    3,  4,  4,  6,  7,  8, 
    3,  3,  4,  4,  6,  7, 
    3,  3,  4,  4,  6,  7, 
    3,  4,  4,  5,  6,  7, 
    3,  3,  4,  4,  6,  7, 
    2,  3,  3,  4,  5,  6, 
    2,  4,  3,  5,  5,  7, 
    2,  3,  4,  4,  5,  6, 
    2,  4,  4,  5,  5,  7, 
    2,  3,  4,  4,  5,  7, 
    2,  3,  4,  5,  5,  7, 
    2,  3,  4,  5,  5,  7, 
    2,  3,  4,  5,  5,  7, 
    2,  3,  3,  4,  5,  7, 
    2,  3,  3,  4,  5,  7, 
    2,  3,  3,  4,  5,  7, 
    2,  3,  3,  4,  5,  7, 
    4,  3,  4,  5,  5,  7 
};

static
void SetProgramName(FileName)
   char *FileName;
{
   char *c_ptr=UtilStrRChr(FileName, '/');

   if (c_ptr != NULL && *(++c_ptr) != '\0') {
      strcpy(progName, c_ptr);
   } else {
      strcpy(progName, TOOL_NAME);
   }
}

static
int PrTgifLoad(FileName)
   char *FileName;
{
   struct ObjRec *obj_ptr;
   char full_name[MAXPATHLENGTH+1], tmp_filename[MAXPATHLENGTH+1];
   int len, read_status;
   FILE *fp;
   int tmp_linenum, obj_ext_len, sym_ext_len;
   char obj_ext_str[MAXSTRING+1], sym_ext_str[MAXSTRING+1];

   sprintf(obj_ext_str, ".%s", OBJ_FILE_EXT);
   obj_ext_len = strlen(obj_ext_str);
   sprintf(sym_ext_str, ".%s", SYM_FILE_EXT);
   sym_ext_len = strlen(sym_ext_str);

   len = strlen(FileName);

   if ((len>=obj_ext_len &&
         strcmp(&FileName[len-obj_ext_len], obj_ext_str) == 0) ||
         (len>=sym_ext_len &&
         strcmp(&FileName[len-sym_ext_len],sym_ext_str)==0)) {
      strcpy(full_name, FileName);
   } else {
      sprintf(full_name, "%s.%s", FileName, OBJ_FILE_EXT);
   }
   if ((fp=fopen(full_name, "r")) == NULL) {
      fprintf(stderr, "Can not open '%s'.\n", full_name);
      return FALSE;
   }

   strcpy(tmp_filename, scanFileName);
   tmp_linenum = scanLineNum;
   strcpy(scanFileName, full_name);
   scanLineNum = 0;

   fprintf(stderr, "Reading '%s' ...\n", full_name);

   while ((read_status = ReadObj(fp, &obj_ptr)) == TRUE) {
      if (obj_ptr != NULL) {
         AdjForOldVersion(obj_ptr);
         AddObj(NULL, topObj, obj_ptr);
      }
   }

   strcpy(scanFileName, tmp_filename);
   scanLineNum = tmp_linenum;

   fclose(fp);

   if (read_status == INVALID) {
      fprintf(stderr, "File version too large (=%1d).  Abort processing!\n",
            fileVersion);
      return FALSE;
   }
   if (cmdLineHasPageNum) {
      if (cmdLinePageNum > lastPageNum) {
         fprintf(stderr, "%s is skipped because it only contains %1d pages!\n",
               FileName, lastPageNum);
         return FALSE;
      }
      GotoPageNum(cmdLinePageNum);
   }
   return TRUE;
}

static
void DoPrTgif(argc, argv, from_prtgif)
   int argc, from_prtgif;
   char *argv[];
{
   char inbuf[MAXSTRING+1];
   char *c_ptr;
   int len, argc_to_be;

   if ((argc_to_be=ProcessPrTgifOptions(argc, argv, from_prtgif)) == INVALID) {
      return;
   }
   while (argc > argc_to_be) {
      argc--; argv++;
   }
   InitPaperSize();
   if (argc <= 0) {
      fprintf(stderr, "\nTgif File Name to Print> ");
      fflush(stderr);
      while (fgets(inbuf, MAXSTRING, stdin) != NULL) {
         len = strlen(inbuf);
         if (len > 0) {
            if (inbuf[--len] == '\n') inbuf[len] = '\0';
            if (PrTgifLoad(inbuf)) {
               colorDump = FALSE;
               if (cmdLineOneFilePerPage) {
                  if (pageLayoutMode == PAGE_TILE) {
                     fprintf(stderr, "%s %s.\n",
                           "Cannot use -one_file_per_page in",
                           "TILED page mode");
                  } else {
                     if (cmdLineHasPageNum) {
                        sprintf(cmdLinePageNumStr, "%1d", cmdLinePageNum);
                        GotoPageNum(cmdLinePageNum);
                        Dump(*argv);
                     } else {
                        cmdLineHasPageNum = TRUE;
                        for (cmdLinePageNum=1; cmdLinePageNum <= lastPageNum;
                              cmdLinePageNum++) {
                           sprintf(cmdLinePageNumStr, "%1d", cmdLinePageNum);
                           GotoPageNum(cmdLinePageNum);
                           Dump(inbuf);
                        }
                        cmdLineHasPageNum = FALSE;
                     }
                  }
               } else {
                  Dump(inbuf);
               }
               DelAllObj();
            }
         }
         fprintf(stderr, "\nTgif File Name to Print> ");
         fflush(stderr);
      }
   } else {
      for ( ; argc > 0; argc--, argv++) {
         lastFile = (argc == 1);
         if (PrTgifLoad(*argv)) {
            colorDump = FALSE;
            if (cmdLineOneFilePerPage) {
               if (pageLayoutMode == PAGE_TILE) {
                  fprintf(stderr, "%s %s.\n",
                        "Cannot use -one_file_per_page in",
                        "TILED page mode");
               } else {
                  if (cmdLineHasPageNum) {
                     sprintf(cmdLinePageNumStr, "%1d", cmdLinePageNum);
                     GotoPageNum(cmdLinePageNum);
                     Dump(*argv);
                  } else {
                     cmdLineHasPageNum = TRUE;
                     for (cmdLinePageNum=1; cmdLinePageNum <= lastPageNum;
                           cmdLinePageNum++) {
                        sprintf(cmdLinePageNumStr, "%1d", cmdLinePageNum);
                        GotoPageNum(cmdLinePageNum);
                        Dump(*argv);
                     }
                     cmdLineHasPageNum = FALSE;
                  }
               }
            } else {
               Dump(*argv);
            }
            DelAllObj();
         }
      }
   }
   CleanUpPaperSize();
}

/*
 * static
 * void Prompt2 (PromptStr, OpName, FileName)
 *    char	* PromptStr, * OpName, * FileName;
 * {
 *    char	inbuf[80];
 * 
 *    printf (PromptStr);
 *    fgets (inbuf, 80, stdin);
 *    sscanf (inbuf, "%s%s", OpName, FileName);
 * }
 *
 * static
 * void Prompt3 (PromptStr, AttrName, ColorName, ValName)
 *    char	* PromptStr, * AttrName, * ColorName, * ValName;
 * {
 *    char	inbuf[80];
 * 
 *    printf (PromptStr);
 *    fgets (inbuf, 80, stdin);
 *    sscanf (inbuf, "%s%s%s", AttrName, ColorName, ValName);
 * }
 */

static
void DoTgif(argc, argv)
   int argc;
   char *argv[];
   /* All these strangeness with strings are related to */
   /*    Prolog's foreign function interface. */
{
   register int i;
   char op_name[80], file_name[80];
   char *sp[6], *func_strp;
/*
 * char	color_name[80], val_name[80];
 * char	attr_name[80], speed_name[80], id_name[80];
 */

   if (!ProcessTgifOptions(argc, argv, file_name)) return;

   if (file_name[0] == '\0') {
      MainLoop("init", "", &func_strp,
            &sp[0], &sp[1], &sp[2], &sp[3], &sp[4], &sp[5]);
   } else {
      MainLoop("init", file_name, &func_strp,
            &sp[0], &sp[1], &sp[2], &sp[3], &sp[4], &sp[5]);
   }
/*
 * for (i = 0; i < 6; i++)
 *    if (strcmp (sp[i], "") != 0)
 *       printf ("%s ", sp[i]);
 *    else
 *       break;
 * printf ("\n");
 */

   while (TRUE) {
      char s[80];

      strcpy(s, func_strp);
      s[4] = '\0';

      DeallocStrings(&func_strp,&sp[0],&sp[1],&sp[2],&sp[3],&sp[4],&sp[5]);

      if (strcmp(s, "Quit") == 0) {
         *file_name = '\0';
         MainLoop("quit", file_name, &func_strp,
               &sp[0], &sp[1], &sp[2], &sp[3], &sp[4], &sp[5]);
         DeallocStrings(&func_strp,&sp[0],&sp[1],&sp[2],&sp[3],&sp[4],&sp[5]);
         break;
      }

      Msg("Returned from basic driver.");
/*
 *    Prompt2("Input an operation and a sub command.\n",op_name,file_name);
 *
 *    if (strcmp(op_name, "animate") == 0)
 *    {
 *       Prompt3("Input poly_id, speed, color.\n", id_name, speed_name,
 *             color_name);
 *       Animate(file_name, id_name, speed_name, color_name, &func_strp);
 *       printf("Animate RETURNs --> %s %s %s\n", func_strp, sp[0], sp[1]);
 *    }
 *    if (strcmp(op_name, "upd_attr_val") == 0)
 *    {
 *       Prompt3("Input attrname, color and value.\n", attr_name, color_name,
 *             val_name);
 *       UpdAttrVal(file_name, attr_name, color_name, val_name, &func_strp);
 *       printf("UpdAttrVal RETURNs --> %s %s %s\n", func_strp, sp[0], sp[1]);
 *    }
 */
      *op_name = *file_name = '\0';
      MainLoop(op_name, file_name, &func_strp,
            &sp[0], &sp[1], &sp[2], &sp[3], &sp[4], &sp[5]);
/*
 *    printf("RETURN --> %s ", func_strp);
 *    for (i = 0; i < 6; i++)
 *       if (strcmp(sp[i], "") != 0)
 *          printf("%s ", sp[i]);
 *       else
 *          break;
 *    printf("\n");
 */
   }
}

int main(argc, argv)
   int argc;
   char *argv[];
{
   register int i, from_prtgif;

   origArgC = argc;
   origArgV = argv;

   PRTGIF = FALSE;
   from_prtgif = FALSE;
   SetProgramName(*argv);
   for (i = 1; i < argc; i++) {
      if (strcmp(argv[i],"-prtgif")==0) {
         PRTGIF = TRUE;
         from_prtgif = TRUE;
         break;
      } else if ((strcmp(argv[i],"-print")==0) ||
            (strcmp(argv[i],"-PRINT")==0)) {
         PRTGIF = TRUE;
         break;
      }
   }
/*
 * malloc_debug (1);
 */
   if (PRTGIF) {
      DoPrTgif(argc, argv, from_prtgif);
   } else {
      DoTgif(argc, argv);
   }
   return 0;
}
